### ISSUE TITLE
IncompleteOperationErrors are imprecise

### ISSUE THREAD

**Issue by johspaeth** [CROSSINGTUD/CryptoAnalysis#51]
The analysis generates a lot of [IncompleteOperationErrors](https://github.com/CROSSINGTUD/CryptoAnalysis/blob/master/CryptoAnalysis/src/main/java/crypto/analysis/errors/IncompleteOperationError.java). 

Extend the test cases found [here](https://github.com/CROSSINGTUD/CryptoAnalysis/blob/master/CryptoAnalysis/src/test/java/test/headless/HeadlessTests.java) to test more thoroughly for these errors.

**Comment by mbruns42** [comment]
Just to be sure, the current state is 
- All tests pass except the HeadlessTests 
- oracleExample fails because of crypto.analysis.errors.RequiredPredicateError in method <Crypto.PWHasher: java.lang.Boolean verifyPWHash)
- cogniCryptDemoExample fails because of ImpreciseValueExtractionError in method <example.ImpreciseValueExtractionErrorExample: void main(java.lang.String[])>

Is that correct?

**Comment by johspaeth** [comment]
That is indeed weird, especially because shippable told that the Build suceeded.
On my machine, the tests failed as well and I have just pushed the fixes to the repo. You should be fine, if you pull.

**Comment by mbruns42** [comment]
Works for me too, now. 
How does the compilation of [OracleExample/src/main/Main.java](https://github.com/CROSSINGTUD/CryptoAnalysis/blob/master/CryptoAnalysisTargets/OracleExample/src/main/Main.java) work? It seems to be configured in Eclipse somehow, but I'm not sure how to start the correct one.

**Comment by johspaeth** [comment]
The class is just compiled to the `bin` folder (of the Eclipse project). This [test case](https://github.com/CROSSINGTUD/CryptoAnalysis/blob/cbc11580a83b80c1d53a608411567ad13bcb38d7/CryptoAnalysis/src/test/java/test/headless/HeadlessTests.java#L67) then executes the analysis on the class files contained in the directory.

**Comment by mbruns42** [comment]
Please add examples where too many of these errors occur.

**Comment by johspaeth** [comment]
Have a look at the CogniCrypt reports which contain a couple of IncompleteOperationErrors. 

**at.favre.lib-hkdf-1.0.0**
- [CogniCrypt Report](https://gist.github.com/johspaeth/af18f8fe4f2d47c0cc22504cfbdc1b3d).  
- [Source Code](https://github.com/patrickfav/hkdf/tree/master/src/main/java/at/favre/lib/crypto).

**com.amazonaws-amazon-kinesis-aggregator-1.0.3**
- [CogniCrypt Report](https://gist.github.com/johspaeth/569341cf1f41bc09f681eb90231bf9ad)
- [Source Code](https://github.com/awslabs/kinesis-aggregation/blob/master/java/KinesisAggregator/src/main/java/com/amazonaws/kinesis/agg/AggRecord.java)

If you classify the findings as false positives, it would be great to understand why the analysis reports the issues.

**Comment by mbruns42** [comment]
Here is the analysis of these errors: [IncompleteOperationError.pdf](https://github.com/CROSSINGTUD/CryptoAnalysis/files/2310977/IncompleteOperationError.pdf)

TL;DR: 
- One potential bug when calling doFinal(), will be investigated
 - Improvement: Add parameters to error message when expected call and observed call have the same name
- False positives may result from libraries initializing crypto objects, but only performing remaining operations when API functions are called
- Improvement suggestion: Number of displayed errors could be reduced by not reporting same IncompleteOperationError for different objects

**Comment by johspaeth** [comment]
Great. I think we should actually have a wiki page explaining the false positives due to common over-approximation. (Loops, Library Analysis, etc).
Do you want to write a wiki about these?

**Comment by mbruns42** [comment]
I'd be happy to add these ones, but I think others have a better overview and may need to complete the wiki.

**Comment by mbruns42** [comment]
Potential bug seems to be weird handling of loops. When the beginning of the loop separates one method in the correct sequence from the next, we get two `IncompleteOperationExceptions`, one in the loop, one outside. 

Can be observed in `useDoFinalInLoop` in #74  in [CryptoAnalysisTargets/OracleExample/src/main/Main.java](https://github.com/CROSSINGTUD/CryptoAnalysis/pull/74/files#diff-0dbcba8f2534d0bcb0aa6084d6a9f2e1)

Can/should we do something about that?

**Comment by johspaeth** [comment]
I had a look at it, I wondered what we can actually do about it. It is related to [this issue](https://github.com/CROSSINGTUD/CogniCrypt/issues/121)

The same also happens in that simpler case:

```
foo(){
  Cipher c= Cipher.getInstance("AES");
  c.init(Cipher.ENCRYPT_MODE, key);
  if(staticallyUnknown()){
     c.doFinal("".getBytes());
  }
  return;
}
```
If the program executes along the false branch, the `doFinal` call is missing. Hence, the analysis should report one error. It reports two as of the following logic:

1. The analysis detects at the `return` statement, that variable `c` does not escape `foo()`. :+1: 
2. At the `return` statement. the variable's object is not in an accepting. :+1: 
3. From the `return` statement the analysis [computes](https://github.com/CROSSINGTUD/WPDS/blob/eb235a35bc5809c1df51f7f32f875ec02ae55079/boomerangPDS/src/main/java/boomerang/results/ForwardBoomerangResults.java#L89) the last statement(s) that use the variable `c` and report the error(s) at those statements. In the example the backward control-flow reachable statement using `c` are the `doDinal` and the `init` calls. :-1:

**Comment by mbruns42** [comment]
So we should decide whether to throw IncompleteOperationErrors EITHER on the return statement or on the destructing statement?
Would choosing just one lead to false negatives?

**Comment by johspaeth** [comment]
No, we would [like to report](https://github.com/CROSSINGTUD/CogniCrypt/issues/121) on the last statement _using_ the variable. It is strange to get a report at the return statement as it does not even use variabel `c` and their might be many lines between the last usage of the variable and the return.

**Comment by mbruns42** [comment]
Ok, that makes sense. So how should we proceed?
